### DISCUSSION TITLE
Beyond the Filesystem: a More Complete Emulator

### DISCUSSION THREAD

**Discussion by james-pre** [zen-fs/core#106] [Ideas]
A very common theme I've seen is users of ZenFS looking to go beyond just emulating the filesystem. Both @chinonso098 and @anthonyhoegberg are looking to develop graphical pseudo-OSes.

I've also been interested and would like to know if anyone would be interested in collaborating on a more complete system emulator. I would love to work on a pseudo-kernel, written in Typescript, but creating even part of a complete operating system is much more complicated.

**Comment by chinonso098** [comment]
@james-pre  I like the sound of that. A kernel in a considerably more complex and ambitious project. I am open to collaborating on such a project.

**Comment by anthonyhoegberg** [comment]
A pseudo-kernel, sounds quite nice, I was wondering what you had in mind exactly?

With ZenFS the filesystem part is basically done, and my WebOS is basically just an universal window manager right now that manages iframes and does window resizing docking etc. Detects clicks on the iframe (cross origin ones too without permission) and manages the windows basically.

I was planning on expanding on that, keeping each part as its own module to separate code.

Would a kernel have a bunch of parts like fs, ui, backend, communication protocols etc?

**Comment by james-pre** [comment]
The kernel would serve as the glue holding everything together- FS, Inter-process communication, permissions, "device" drivers, task scheduling, etc.

Since it would be based in Typescript, memory management isn't really a concern.

I think it would be cool if it was cross platform.

The kernel would be completely separate from user-mode code (i.e. basically all of the applications).

The UI (graphical and terminal) and most programs are not part of a kernel, but when combined with one serve as a complete operating system.

**Reply by anthonyhoegberg** [reply]
It would make sense to isolate processes etc right? maybe run them in web workers etc?

**Reply by chinonso098** [reply]
@james-pre  @anthonyhoegberg  When do we begin, and what is the game plan ?

**Reply by james-pre** [reply]
If either of you have Discord, we could chat more informally on there? You can [join the server](https://zenfs.dev/discord).

**Reply by anthonyhoegberg** [reply]
I got work rn but im down to talk and discuss this topic after that to see what kinds of benefits this kind of project can have and how such a project could be made if feasible.
Our time zones might not match up (currently 10:23 am for me since i live in japan) 
(does zenfs have a discord server or smth?)

**Reply by james-pre** [reply]
ZenFS used to use a discord server, but I thought it would be better to use GitHub discussions. Between us though, and working on this project, we will need a lot more communication.

As for the timezones, I'm sure we can figure something out.

**Reply by anthonyhoegberg** [reply]
> ####
> [ anthonyhoegberg](/anthonyhoegberg)

@chinonso098 i messaged james-pre on discord, if you message them too maybe we could talk about the kernel project etc on there the three of us.

**Reply by chinonso098** [reply]
@anthonyhoegberg @james-pre  I'm trying to recover my discord account. They keep sending me verification emails that are not getting to my inbox, spam or deleted folder.

**Reply by chinonso098** [reply]
I’m now able to access my Discord account. Username: nahaelem

**Comment by james-pre** [comment]
@mathiscode Are you interested in collaborating? We have a Discord group chat and [server](https://zenfs.dev/discord), please reach out to me on Discord if you're interested.

**Reply by mathiscode** [reply]
@james-pre Absolutely - joining now!

**Reply by mathiscode** [reply]
@james-pre Ah, unable to accept invite - my username is `mathiscode`
